### PR TITLE
Determine input shapes

### PR DESCRIPTION
**Pull request by l-bat**
```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2019r2.0:16.04
build_image:Custom Win=openvino-2019r2.0
build_image:Custom Mac=openvino-2019r2.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom Win=windows-1
test_opencl:Custom=OFF

buildworker:Custom=linux-1
test_opencl:Custom=ON
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/test/test_ie_models.cpp [hunk 1]
```diff
@@ -319,5 +319,50 @@ INSTANTIATE_TEST_CASE_P(/**/,
...
+    OpenVINOModelTestCaseInfo modelInfo = getOpenVINOTestModels().find("age-gender-recognition-retail-0013")->second;
+
+    std::string modelPath = isFP16 ? modelInfo.modelPathFP16 : modelInfo.modelPathFP32;
+
+    std::string xmlPath = findDataFile(modelPath + ".xml");
+    std::string binPath = findDataFile(modelPath + ".bin");
+
+    // Single Myriad device cannot be shared across multiple processes.
+    if (target == DNN_TARGET_MYRIAD)
+        resetMyriadDevice();
```
Comment (dkurt) [inline modules/dnn/test/test_ie_models.cpp]: Why so? Here is a single process.

File: modules/dnn/test/test_ie_models.cpp [hunk 2]
```diff
@@ -319,5 +319,50 @@ INSTANTIATE_TEST_CASE_P(/**/,
...
+    model.setPreferableTarget(target);
+    model.predict(frame, outs);
+
+    Net net = readNet(xmlPath, binPath);
+    Mat input = blobFromImage(frame, 1.0, Size(62, 62));
+    net.setInput(input);
+    net.setPreferableBackend(DNN_BACKEND_INFERENCE_ENGINE);
+    net.setPreferableTarget(target);
+
+    std::vector<String> outNames = getOutputsNames(net);
```
Comment (dkurt) [inline modules/dnn/test/test_ie_models.cpp]: Please replace to new `getUnconnectedOutLayersNames`

File: modules/dnn/src/model.cpp [hunk 3]
```diff
@@ -100,8 +106,7 @@ void Model::setInputParams(double scale, const Size& size, const Scalar& mean,
 
 void Model::predict(InputArray frame, OutputArrayOfArrays outs)
 {
-    std::vector<Mat> outputs;
-    outs.getMatVector(outputs);
+    std::vector<Mat> & outputs = *(std::vector<Mat> *)outs.getObj();
```
Comment (dkurt) [inline modules/dnn/src/model.cpp]: Do we really need this change?

File: modules/dnn/src/model.cpp [hunk 4]
```diff
@@ -26,7 +26,13 @@ struct Model::Impl
...
     {
         if (size.empty())
-            CV_Error(Error::StsBadSize, "Input size not specified");
+        {
+            std::vector<MatShape> inLayerShapes;
+            std::vector<MatShape> outLayerShapes;
+            net.getLayerShapes(MatShape(), 0, inLayerShapes, outLayerShapes);
+            CV_Assert_N(!inLayerShapes.empty(), inLayerShapes[0].size() == 4);
+            size = Size(inLayerShapes[0][3], inLayerShapes[0][2]);
+        }
```
Comment (dkurt) [inline modules/dnn/src/model.cpp]: I think that we need to set it in the constructor.

File: modules/dnn/src/dnn.cpp [hunk 5]
```diff
@@ -2595,14 +2608,22 @@ Net Net::readFromModelOptimizer(const String& xml, const String& bin)
...
+        inp_shapes.push_back(std::vector<int>(dims.begin(), dims.end()));
     }
 
     Net cvNet;
     cvNet.setInputsNames(inputsNames);
 
+    for (int inp_id = 0; inp_id < inputsNames.size(); ++inp_id)
+    {
+        cvNet.setInput(Mat(inp_shapes[inp_id], CV_32F), inputsNames[inp_id]);
+    }
```
Comment (dkurt) [inline modules/dnn/src/dnn.cpp]: Please add some comment why we set input here.

File: modules/dnn/src/dnn.cpp [hunk 6]
```diff
@@ -2455,6 +2455,19 @@ struct Net::Impl
...
+            CV_Assert(!layers[0].outputBlobs.empty());
+            ShapesVec shapes;
+            for (int i = 0; i < layers[0].outputBlobs.size(); i++)
+            {
+                Mat& inp = layers[0].outputBlobs[i];
+                CV_Assert(inp.total());
+                shapes.push_back(shape(inp));
+            }
+            inOutShapes[0].in = shapes;
+         }
```
Comment (dkurt) [inline modules/dnn/src/dnn.cpp]: Can you briefly describe what this part of code does?

File: modules/dnn/src/model.cpp [hunk 7]
```diff
@@ -45,6 +42,15 @@ Model::Model(const String& model, const String& config)
     : Net(readNet(model, config)), impl(new Impl)
 {
     impl->outNames = getUnconnectedOutLayersNames();
+
+    if (impl->size.empty())
```
Comment (dkurt) [inline modules/dnn/src/model.cpp]: This is constructor. The `size` field is empty

File: modules/dnn/src/model.cpp [hunk 8]
```diff
@@ -45,6 +42,15 @@ Model::Model(const String& model, const String& config)
...
+    {
+        std::vector<MatShape> inLayerShapes;
+        std::vector<MatShape> outLayerShapes;
+        getLayerShapes(MatShape(), 0, inLayerShapes, outLayerShapes);
+        if (!inLayerShapes.empty() && inLayerShapes[0].size() == 4)
+            impl->size = Size(inLayerShapes[0][3], inLayerShapes[0][2]);
```
Comment (dkurt) [inline modules/dnn/src/model.cpp]: Please add the same logic to the second constructor

File: modules/dnn/src/model.cpp [hunk 9]
```diff
@@ -25,9 +25,6 @@ struct Model::Impl
 
     void predict(Net& net, const Mat& frame, std::vector<Mat>& outs)
     {
-        if (size.empty())
-            CV_Error(Error::StsBadSize, "Input size not specified");
```
Comment (dkurt) [inline modules/dnn/src/model.cpp]: I think that we need to keep this check.

File: modules/dnn/test/test_ie_models.cpp [hunk 10]
```diff
@@ -319,5 +305,46 @@ INSTANTIATE_TEST_CASE_P(/**/,
...
+    net.forward(refs, outNames);
+
+    CV_Assert(refs.size() == outs.size());
+    for (int i = 0; i < refs.size(); ++i)
+        normAssert(outs[i], refs[i]);
+}
+
+INSTANTIATE_TEST_CASE_P(/**/,
+    DNNTestHighLevelAPI, testing::ValuesIn(getAvailableTargets(DNN_BACKEND_INFERENCE_ENGINE))
+);
```
Comment (dkurt) [inline modules/dnn/test/test_ie_models.cpp]: I think it'd be nice to add a test for Caffe as well (or modify one of existing).

File: modules/dnn/src/model.cpp (hunk 3 above)
Comment (dkurt) [reply modules/dnn/src/model.cpp]: Please try this: https://github.com/opencv/opencv/pull/15256/files#diff-6de938636906f03a3c475c556f5c8878R105
and https://github.com/opencv/opencv/pull/15256/files#diff-6de938636906f03a3c475c556f5c8878R26

**Comment by dkurt** [comment]
@l-bat, May I ask you to rebase this branch? Thanks!

**Comment by dkurt** [comment]
And please include this line back as well in this PR: https://github.com/opencv/opencv/pull/15184/files#diff-060059506235351e720719798d655782L308

File: modules/dnn/src/dnn.cpp [hunk 11]
```diff
@@ -2595,14 +2607,23 @@ Net Net::readFromModelOptimizer(const String& xml, const String& bin)
...
+        inp_shapes.push_back(std::vector<int>(dims.begin(), dims.end()));
     }
 
     Net cvNet;
     cvNet.setInputsNames(inputsNames);
 
+    // set empty input to determine input shapes
+    for (int inp_id = 0; inp_id < inputsNames.size(); ++inp_id)
+    {
+        cvNet.setInput(Mat(inp_shapes[inp_id], CV_32F), inputsNames[inp_id]);
```
Comment (alalek) [inline modules/dnn/src/dnn.cpp]: @l-bat @dkurt Looks like this patch is landed for "master" branch only. What is about backporting of this PR?

Comment (dkurt) [reply modules/dnn/src/dnn.cpp]: I do not mind. Besides `cv::dnn::Model` other code should be compatible. Let me try to port it.

**Pull request merged**